analysis/src/vm/function_desugar.rs

```diff
@@ -0,0 +1,207 @@
+//! Desugar asm functions which operate on values to functions which operate on witness columns
+//! TODO: Using the same term "function" for both is a bit awkward
+
+use ast::{
+    asm_analysis::{
+        utils::previsit_expression_mut, AnalysisASMFile, AssignmentStatement, FunctionStatement,
+        FunctionStatements, InstructionStatement, Machine,
+    },
+    parsed::{
+        asm::{Param, ParamList},
+        Expression,
+    },
+};
+use number::FieldElement;
+use std::{collections::HashMap, marker::PhantomData, ops::ControlFlow};
+
+use crate::utils::{parse_instruction_definition, parse_register_declaration};
+
+/// Desugar VM functions
+pub fn desugar<T: FieldElement>(file: AnalysisASMFile<T>) -> AnalysisASMFile<T> {
+    FunctionDesugar::default().generate(file)
+}
+
+#[derive(Default)]
+struct FunctionDesugar<T> {
+    marker: PhantomData<T>,
+}
+
+fn input_at(i: usize) -> String {
+    format!("_input_{}", i)
+}
+fn output_at(i: usize) -> String {
+    format!("_output_{}", i)
+}
+
+fn substitute<T>(mut e: Expression<T>, inputs: &HashMap<String, usize>) -> Expression<T> {
+    previsit_expression_mut(&mut e, &mut |e| {
+        if let Expression::PolynomialReference(r) = e {
+            if let Some(v) = inputs.get(&r.name).map(|i| input_at(*i)) {
+                r.name = v;
+            }
+        };
+        ControlFlow::Continue::<()>(())
+    });
+    e
+}
+
+impl<T: FieldElement> FunctionDesugar<T> {
+    fn generate(&self, file: AnalysisASMFile<T>) -> AnalysisASMFile<T> {
+        AnalysisASMFile {
+            machines: file
+                .machines
+                .into_iter()
+                .map(|(name, m)| (name, self.generate_machine_rom(m)))
+                .collect(),
+        }
+    }
+
+    fn generate_machine_rom(&self, mut machine: Machine<T>) -> Machine<T> {
+        if machine.has_pc() {
+            let latch = "instr_return";
```

Comment:
Right! So the only user-facing one is `return`, so it could be banned in the parser or type checker (any preference @leonardoalt?). The other ones start with `_` which the user is not expected to use. This could be enforced as well.

Reply:
yea fine with either parser or type checker

Reply:
Added to type checker